### PR TITLE
Refactors binterp in modal_aero_opt.F90

### PR DESCRIPTION
**Pull request by singhbalwinder**
Refactors binterp by factoring out a routine with repeated calculations. Renames variables add some comments.

### CODE REVIEW

**Review by jeff-cohere — APPROVED** [review]
Thanks, Balwinder!

File: components/eam/src/physics/cam/modal_aer_opt.F90 [hunk 1]
```diff
@@ -1553,71 +1551,114 @@ end subroutine modal_size_parameters
...
+        dx     = (ref_table(ip1)-ref_table(ix(ic)))
+        if(abs(dx) > threshold) then
+           tt(ic) = (ref_ind(ic)-ref_table(ix(ic)))/dx
+           if (present(do_print) .and. do_print) then
+              if(tt(ic) < 0._r8 .or. tt(ic) > 1._r8) then
+                 write(iulog,*) 'tt,ref_ind,ix,ref_table,dx=',tt(ic),ref_ind(ic),ix(ic),ref_table(ix(ic)),dx
+              endif
+           endif
+        else
+           tt(ic) = 0._r8
```
Comment (tangshuaiqi) [inline components/eam/src/physics/cam/modal_aer_opt.F90]: can we initiate tt(:) = 0._r8 at the beginning so we don't need the else condition here and below? 
It looks weird that all elements of tt (and likely ix) are overwritten here, why does it need to be an "inout" variable? Also they are not used in anywhere else in modal_aero_sw and modal_aero_lw. Can it be put as local variable in binterp?

File: components/eam/src/physics/cam/modal_aer_opt.F90 [hunk 2]
```diff
@@ -1553,71 +1551,114 @@ end subroutine modal_size_parameters
...
+
+  !local
+  integer  :: ip1, jp1, icoef, ic
+  real(r8) :: tu(pcols), tuc(pcols),tcu(pcols), tcuc(pcols)
+
+  if(itab(1) <= 0) then
+
+     !compute factors for the real part
+     call compute_factors(prefr, ncol, ref_real, ref_real_tab, & !in
+          itab, ttab) !out
```
Comment (tangshuaiqi) [inline components/eam/src/physics/cam/modal_aer_opt.F90:1668]: this should be inout

**Review by tangshuaiqi — APPROVED** [review]
looks good to me. Only have a minor correction and a comment to discuss ttab (or tt).

File: components/eam/src/physics/cam/modal_aer_opt.F90 (hunk 2 above)
Comment (singhbalwinder) [reply components/eam/src/physics/cam/modal_aer_opt.F90:1668]: I ended up keeping it `out` following your other comment.

File: components/eam/src/physics/cam/modal_aer_opt.F90 (hunk 1 above)
Comment (singhbalwinder) [reply components/eam/src/physics/cam/modal_aer_opt.F90]: That's a good observation. I have removed both the `else` statements and simplified the code. I have also make `tt` and `ix` intent-out (as you suggested).
Both `tt` and `ix` (real and imaginary counterparts) are computed in the first call of binterp and used in the subsequent calls. We can make them local if we just had one binterp call (such as in long wave). For shortwave, we make three calls, so we need them for the second and the third call as input.

Another option would be to declare them with `save` or declare them globally at the module level. Both these solutions can cause issues with threading. So sending them via arguments is the best solution and it also makes it clear what the dependencies are by just looking at the argument list.

**Review by brigaudet — APPROVED** [review]
Looks great to me!

**Comment by singhbalwinder** [comment]
All tests passed:
$ /compyfs/sing201/e3sm_scratch/tmp_biinterp/cs.status.comp_tmp_biinterp_mam4_org_v2_baselines -s
comp_tmp_biinterp_mam4_org_v2_baselines: 9 tests
  PASS ERP_P64x1.ne4pg2_oQU480.F2010.compy_intel
  PASS ERS.ne4pg2_oQU480.F2010.compy_intel
  PASS PEM_P64x1.ne4pg2_oQU480.F2010.compy_intel
  PASS PET.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_D_Ln5_P1x1.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_D_Ln5_P32x1.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_D.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_Ly1.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS.ne30pg2_EC30to60E2r2.F2010.compy_intel

**Pull request merged**